src/openapi-generator.ts

```diff
@@ -346,87 +363,105 @@ export class OpenAPIGenerator {
     };
   }
 
+  private generateSchemaWithMetadata<T>(zodSchema: ZodSchema<T>) {
+    const innerSchema = this.unwrapChained(zodSchema);
+    const metadata = this.getMetadata(zodSchema);
+    const defaultValue = this.getDefaultValue(zodSchema);
+
+    const result = metadata?.metadata?.type
+      ? { type: metadata?.metadata.type }
+      : this.toOpenAPISchema(innerSchema, zodSchema.isNullable(), defaultValue);
+
+    return metadata?.metadata
+      ? this.applySchemaMetadata(result, metadata.metadata)
+      : omitBy(result, isNil);
+  }
+
   /**
    * Generates an OpenAPI SchemaObject or a ReferenceObject with all the provided metadata applied
    */
   private generateSimpleSchema<T>(
```

Comment:
Idea for a refactoring at some point (not now) - currently I believe we do a lot of duplicated unwrapping, for example we call `unwrapChained` at multiple points on the same schema. It may be possible to get some good performance gains by reorganizing so that these lookups are cached or reused in some other way.